src/middleware/permissions.js

```diff
@@ -0,0 +1,84 @@
+const url = require('url')
+const boom = require('boom')
+const roles = require('../constants/roles.js')
+const { auth, messages: errMessages } = require('../constants/errors.json')
+const { rejectIfNull } = require('../db/utils')
+
+const Event = require('../models/Event')
+const Place = require('../models/Place')
+const Product = require('../models/Product')
+const User = require('../models/User')
+
+const hasSufficientRole = ({ minRole }) => user => {
+  const orderedRoles = [roles.BASIC, roles.ADMIN, roles.SUPER]
+
+  // throw error on initialisation if passed option is not valid
+  if (!orderedRoles.includes(minRole)) {
+    throw boom.badImplementation()
+  }
+
+  return orderedRoles.indexOf(user.role) >= orderedRoles.indexOf(minRole)
+}
+
+const checkUserOwnsResource = resourceType => resourceId => user => {
+  if (resourceType === User) {
+    return resourceId === user.id.toString()
+      ? Promise.resolve()
+      : Promise.reject(boom.unauthorized(auth.UNAUTHORIZED))
+  }
+
+  return resourceType.findById(resourceId)
+  .then(rejectIfNull(errMessages.GET_ID_NOT_FOUND))
+  // currently using toString(), may be better to use <mongoid>.equals(<mongoid>)
+  .then(doc => doc.owner.toString() === user.id.toString()
+    ? Promise.resolve()
+    : Promise.reject(boom.unauthorized(auth.UNAUTHORIZED))
+  )
+}
+
+const getResourceType = req => {
+  // currently needs url to be of the form '/<resource>/:id'
+  const routeResourceMapping = {
+    events: Event,
+    places: Place,
+    products: Product,
+    users: User
+  }
+  const pathName = url.parse(req.url).pathname
+  return routeResourceMapping[pathName.split('/')[1]]
+}
+
+module.exports =
+  ({ minRole, ownerIsPermitted }) => (req, res, next) => {
+    const resourceType = getResourceType(req)
+
+    if (ownerIsPermitted && !resourceType) {
+      next(boom.badImplementation())
+    }
+
+    if (!req.user || !req.user.id) {
+      return next(boom.unauthorized(auth.UNAUTHORIZED))
+    }
+
+    else if (hasSufficientRole({ minRole })(req.user)) {
+      return next()
```

Comment:
This still makes no sense to me.
This line here says the user is allowed to access the resource, but at this point it is possible that the user is trying to access a resource that they are not allowed, as some of the checks have not been made.

```js
if (userFailsFirstCheck) unautherised
if (userFailsSomeOtherCheck) unautherised
if (somethingElse) unautherised

// authorised!!
```

If you think I am wrong, then ill merge this tomorrow morning.

Reply:
Which checks haven't been made?

Reply:
okay, I see what is going on now (https://github.com/foundersandcoders/open-tourism-platform/issues/134#issuecomment-330338072)